### ISSUE TITLE
questions on merging otf models and segmentation fault

### ISSUE THREAD

**Issue by eric-yu-zhu** [mir-group/flare_pp#49]
Dear flare team,

I am trying to reproduce your recent work Active learning of reactive Bayesian force fields applied to heterogeneous catalysis dynamics of H/Pt. Nat Commun 13, 5183 (2022). Thank you so much for sharing data and code in addition to presenting the results.

Below are some questions:

1. When the four otf models (H2, Pt-bulk, PtH, Pt-surface) are trained independently, how do I "pool together all the structures and sparse environments" to "construct the final SGP model"? To be specific, how PtH2.txt is created given H2.out, Pt.out, PtH.out, and Pt.out? Is it done with flare or I need to write a script myself?

2. I ran into a segmentation fault with h2_otf.py. It is noticed that the first element of the likelihood gradient is an Nan which triggered the error. 
    Likelihood gradient:
    [            nan  1.36049602e+01 -2.66814500e+03  1.40834906e+04]
    Likelihood:
    310.80666136976754
Further debugging indicates that Kuu_i.inverse() in sparse_gp.cpp/compute_likelihood_stable is problematic. Because of the short cutoff, the two H atoms of a hydrogen molecule can only see each other and hence have identical chemical environments. As a result, the two Kuu rows associated with the two H atoms are identical, making Kuu not invertible. Do you have the same issue in your calculations?

3. What is "single_atom_energies" of SGP_Wrapper? How do I set this parameter?

Thanks,
Eric

**Comment by YuuuXie** [comment]
Hi @eric-yu-zhu ,

1. You can extract DFT frames in all .out files (using `flare.io.otf_parser`, and remember to convert the stress unit from Gpa to eV/A3), and concatenate them and do one offline training to get a final model (see flare colab tutorial). You can also write a script which is equivalent
2. Yes your observation is correct, that the H atoms will likely have identical local environments and make the kernel matrix linear correlated. One thing you can try is to add only one atom each DFT call during the active learning. The other thing you can try is to set up `my_sgp_wrapper.sparse_gp.Kuu_jitter = 1e-4`, the default value is 1e-8. And this value basically adds a small number to the diagonal of Kuu to alleviate the numerical issue.
3. `single_atom_energies` sets up one constant energy for each chemical species, then the actual energy label that SGP takes will be the DFT energy minus the total constant shift. We sometimes set up the value to be the DFT energy of an isolated atom of a certain species. But setting to 0 is also common and seems to work as well.

**Comment by eric-yu-zhu** [comment]
Hi Yu, 

Thanks a lot for the prompt reply!

Q2 and Q3 are clear now. I want to confirm my understanding regarding Q1. What I need to do is to read and pool dft frames (marked by *-Frame) from the out-files (H2.out, Pt.out, PtH.out, Pt.out) to create a dft dataset and then run an offline training. What is used in the final md is the offline trained model. The four trained otf models are "abandoned" after generating the dft dataset. Correct?

Furthermore, here are a few follow-up questions:

4. Why do you only use dft frames in the offline training? How about other MD frames?
 
5. The four otf models have very different atomic structures, and the merged dft dataset will be a hybrid one with varying atom numbers and species. I would assume the "Fake MD" is learning in terms of central atom and its local environment. For this reason, it is capable of training with hybrid systems. Correct? 

6. How do you pool from the four otf models? Random sampling or simply collecting all dft frames? Does the order of frames matter?

7. In addition to the unit conversion, do I need to change the sign or the order of stress tensor components?

Eric

**Comment by YuuuXie** [comment]
Hi @eric-yu-zhu ,
 > What I need to do is to read and pool dft frames (marked by *-Frame) from the out-files (H2.out, Pt.out, PtH.out, Pt.out) to create a dft dataset and then run an offline training. What is used in the final md is the offline trained model. The four trained otf models are "abandoned" after generating the dft dataset. Correct?

Yes. For reading and pooling you can use `otf_parser`

For your other questions
4. Only the DFT data are the one that you use to train the machine learning force field. The MD frames are generated by MLFF itself
5. Yes
6. It will still be an on-the-fly flavor, it starts with the first frame of all your DFT frames, and will go through frames one by one and add high uncertainty frame and atomic envs on the fly. The order of frames matters, similar to the training of neural networks, changing the order of training data definitely results in a different model. I would suggest you to put simple and low temperature configurations first, and append frames with increasing complexity and more extreme conditions.
7. If you are going to save the all dft frames into an xyz file for example (and then feed the xyz file into the offline training), then you need to change the sign, and convert it into a 3x3=9 component array

**Comment by eric-yu-zhu** [comment]
Got it, thanks a lot!

Eric
